src/blueapi/config.py

```diff
@@ -7,10 +8,12 @@
 from bluesky_stomp.models import BasicAuthentication
 from pydantic import BaseModel, Field, TypeAdapter, ValidationError
 
-from blueapi.utils import BlueapiBaseModel, InvalidConfigError
+from blueapi.utils import BlueapiBaseModel, InvalidConfigError, format_errors
 
 LogLevel = Literal["NOTSET", "DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]
 
+ENV_PREFIX = "BLUEAPI."
```

Comment:
This should be `BLUEAPI_`. You can't easily set environment variables containing '.' from the shell which makes this a pain.

Reply:
hmmm, that was in the acceptance criteria

Reply:
Yep, no objection

Reply:
Both python variable names and environment variables are only guaranteed to support `[A-Za-z_][A-Za-z0-9_]?`, which means that trying to parse nested config is a nuisance...
Currently it builds the complete `dict[str, str]` before ever trying to parse it as a Config object, so we can't check the keys of the current object. May have to do that, or limit the config from having keys with an underscore.
`WorkerEventConfig` and `ScratchRepository` already break that.

Reply:
I'm not sure what is your conclusion here @DiamondJoseph

Reply:
Conclusion is that underscores shouldn't work either, @tpoliaw are you happy with just doing

```shell
export "BLUEAPI.foo.bar=baz"
```

...with the quotes, from the command line?

Reply:
Not really...

````sh
$ export "BLUEAPI.foo.bar=baz"
export: not valid in this context: BLUEAPI.foo.bar

Reply:
```sh
env 'BLUEAPI.config.foo=fizz' 'BLUEAPI.config.foo.bar=3' blueapi serve
```

Is the only way I've found so far which isn't great for discoverability.

Reply:
That's annoying, there are other schemes we could use I guess but they obfuscate the truth a bit more

Reply:
@DominicOram as the original issue author how do you feel about dots and underscores?
https://github.com/DiamondLightSource/blueapi/issues/495#issuecomment-2342885860

Reply:
I agree with @tpoliaw that we should be able to set it from the shell. Otherwise I don't have a preference as long as it's documented

Reply:
I'm actually a fan of

```
env 'BLUEAPI.config.foo=bar' blueapi serve
```

If we put it in the `--help` output, is any less discoverable than

```
BLUEAPI.config.foo=bar blueapi serve
```

Reply:
> If we put it in the --help output, is any less discoverable than

were you meaning to put a question mark at the end here?

Reply:
Yes, sorry

Reply:
is it picking up ENV values from the environment though?

Reply:
Should be, but I can't even echo it from the terminal

```
[vid18871@diamrl6120-linux ~]$ env 'BLUEAPI.config.foo=bar' echo ${BLUEAPI.config.foo=bar}
bash: ${BLUEAPI.config.foo=bar}: bad substitution
```

Actually, how about `/` as the delimiter?

Reply:
`/` has the same issue as `.`: `[A-Za-z_][A-Za-z0-9_]?` is the only guaranteed supported (on all (?) shells) regex pattern for both env_vars AND python field names: there is no character we can use as a separator without restricting its use in the python config names. And the only non-alphanumeric character we could use in that case is `_`, which is currently used: it could be removed, but then either all our config is restricted to 1 word or we're doing unpythony things like camelCase.

Reply:
https://stackoverflow.com/a/75709846